cluster_link/replication: add remote data source and partition sink

bharathv opened the pull request:
Wires up the new source and sink with link infrastructure. Adds a couple of new abstractions
partition_data_queue and  `mux_remote_consumer` to decouple the per partition buffering from the main fetch loop.

TODO: need more coverage for the data sink and more test coverage for e2e paths.

Fixes

https://redpandadata.atlassian.net/browse/CORE-10231
https://redpandadata.atlassian.net/browse/CORE-10232
https://redpandadata.atlassian.net/browse/CORE-10233

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
k/tests/dc_fixture: miscenallaneous fixture changes for future use

Mainly creates virtual Start/StopConsumer() methods to make the starting
and stopping optional.

bharathv committed:
k/client/dc: expose subscriptions for testing

bharathv committed:
cluster_link/replication: qualify model usage with :: prefix

There is a conflict with cluster_link::model in future places that
prompts this (somewhat ugly) usage.

bharathv committed:
cluster_link: wireup replication_mgr with link services

Wires up remote_source/local partition sink with link implementaiton.

bharathv committed:
cluster_link: propagate term into link partition changes

Term is needed in the future patches to start/stop replicators API.
plumb an optional term in all places.

bharathv committed:
cluster_link/replication/tests: add test deps implementation

This is for tests so a link test doesn't depend on a remote source. The
source produces random data and sink does some basic accounting of
records.

bharathv committed:
cluster_link/link: make stop() noexcept

Avoid double stop in some cases that may result in double cleanup of
resources. Instead mark stop() noexcept and proceed with link removal.
The onus is in the implementation of stop() to do proper cleanup and not
to throw.

bharathv committed:
cluster_link/replication: partition_data_queue and mux_consumer

Introduces two main abstractions

- partition_data_queue - Isolates data flow per partition.
  Prevents a blocked partition sink from stalling the consumer and
  other partitions.

- mux_remote_consumer - Multiplexes data fetching for all
  partition sinks on a local shard and works with partition_data_queus.
  Detects when a partition queue is not draining, unregisters it to
  allow others to make progress, and re-registers once the partition
  resumes

bharathv committed:
cluster_link/replication: add remote source sink impl

To be wired up in future commits.

bharathv commented:
/dt

bharathv commented:
/dt

bharathv commented:
/dt

bharathv committed:
tests/c: enable cluster linking in DC fixture

bharathv committed:
cluster_linking/tests: add fixture test for partition replicator

bharathv commented:
/dt

bharathv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--777cc09e-6b60-45f7-a19f-c59cd45fff35-->
<!--TEST RESULTS-->
<!--1ec8aab0-19b4-4f5e-9156-293d5c618882-->
<!--TEST RESULTS-->
<!--dd8fe8b6-1d4c-40e4-a154-71703e0e220a-->
<!--TEST RESULTS-->
<!--95b57ee0-d3ea-483b-adb3-d6b5be40d464-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71423">build#71423</a></summary>


| test_class                 | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndShadowIndexingTest | test_reset    | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/71423#0198e8b6-4f63-465f-a49c-80bf0aa9e267 | FLAKY         | 20/21    | upstream reliability is '99.71098265895954'. current run reliability is '95.23809523809523'. drift is 4.47289 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71515">build#71515</a></summary>


| test_class            | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/71515#0198f148-26a0-4350-a37b-883953f7faa5 | FLAKY         | 20/21    | upstream reliability is '99.39148073022312'. current run reliability is '95.23809523809523'. drift is 4.15339 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71572">build#71572</a></summary>


| test_class               | test_method   | test_arguments            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:-------------------------|:--------------|:--------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageScrubberTest | test_scrubber | {"cloud_storage_type": 1} | integration | https://buildkite.com/redpanda/redpanda/builds/71572#0198f320-530d-4fa4-b685-980b9c8e7ce9 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71699">build#71699</a></summary>


| test_class             | test_method                        | test_arguments                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:-----------------------|:-----------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest  | test_migrated_topic_data_integrity | {"include_groups": false, "params": {"cancellation": {"dir": "in", "stage": "prepared"}, "use_alias": true}, "transfer_leadership": true} | integration | https://buildkite.com/redpanda/redpanda/builds/71699#01990638-42e9-48ac-8be8-ce19a001cae1 | FLAKY         | 20/21    | upstream reliability is '99.73544973544973'. current run reliability is '95.23809523809523'. drift is 4.49735 and the allowed drift is set to 50. The test should PASS |
| FetchTest              | fetch_long_poll_test               | null                                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/71699#01990638-42e5-4d75-befc-0a82f87a475b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| RapidTopicRecreateTest | test_topic_rapid_recreation        | null                                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/71699#01990638-42e7-4967-8ded-dd216d1546cc | FLAKY         | 20/21    | upstream reliability is '99.49367088607595'. current run reliability is '95.23809523809523'. drift is 4.25558 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented on src/v/cluster_link/link.cc:
```diff
@@ -89,7 +97,16 @@ ss::future<> link::stop() {
           t->name(),
           _config.name);
-        auto res = co_await stop_task(t.get());
+        auto res_f = co_await ss::coroutine::as_future(stop_task(t.get()));
+        if (res_f.failed()) {
+            vlog(
+              cllog.warn,
+              "Failed to stop task {}: {}",
+              t->name(),
+              res_f.get_exception());
+            continue;
+        }
+        auto res = res_f.get();
```
I am not sure if this is the right approach. What happens if one of th task fails to stop and we continue to shutdown the link, the task can access a nullptr

bharathv replied:
If we don't do it we will have two options

* don't erase the link (memory pile up)
* retry the link stop (double free of resources)

How about making task->stop() as noexcept? The onus will be on tasks to stop cleanly in all cases.

bharathv replied:
> How about making task->stop() as noexcept?

Done.

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,214 @@
+
+void mux_remote_consumer::add(const model::topic_partition& tp) {
+    vlog(cllog.trace, "Adding partition {}", tp);
+    auto holder = _gate.hold();
```
holder is not required here as this code is not async

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,214 @@
+
+void mux_remote_consumer::reset(
+  const model::topic_partition& tp, kafka::offset offset) {
+    auto holder = _gate.hold();
```
gate not needed, if you want to prevent this code from being called after the gate is closed `gate.check()` should be sufficient

bharathv replied:
will do

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,214 @@
+void mux_remote_consumer::add(const model::topic_partition& tp) {
+    vlog(cllog.trace, "Adding partition {}", tp);
+    auto holder = _gate.hold();
+    vassert(!_partitions.contains(tp), "Duplicate partition added: {}", tp);
```
maybe throw, to prevent redpanda crashing if someone made a mistake

bharathv replied:
will do

mmaslankaprv commented on src/v/cluster_link/link.h:
```diff
@@ -45,8 +45,10 @@ class link {
+    ss::future<> handle_on_leadership_change(
+      ::model::ntp ntp,
+      ntp_leader is_ntp_leader,
+      std::optional<::model::term_id>);
```
can you explain why the term is required ?

bharathv replied:
It was originally intended to make start_replicator() idempotent. If start failed we will retry start but meanwhile there could be another start from a leadership change. So having a term handy will help fence stale start requests. Probably over complicated, I feel like it can be removed.

bharathv replied:
After giving it some more thought, I did not do this. Even if we don’t need start_replicator() to be idempotent, we still need a term to trigger a step down in case replicator.start() fails..otherwise, we risk being stuck without a replicator if start throws and there is no leadership change. Given that we need the term anyway, we might as well use it to implement idempotency for the start logic. Lmk if you feel otherwise.

bharathv replied:
added a comment as discussed this morning.

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.h:
```diff
@@ -0,0 +1,95 @@
+     * are add() and reset() (atleast once). add() adds the partition to the
+     * consumer and reset() sets the initial offset for fetching.
+     */
+    ss::future<partition_data_queue::fetch_data>
```
this part seems to be leaking units all the way from `partition_data_queue`, is it possible the queue is destroyed before batches are consumed ? 
If we need to manage total memory usage maybe it would be better to have the per shard sidecar component that would account for all inflight batches memory usage ?

bharathv replied:
> this part seems to be leaking units all the way from partition_data_queue

wdym by leaking? The units are consumed by the [replicator](https://github.com/redpanda-data/redpanda/blob/5b4228f5e1faccb89c238eed62adbf4cdd07120d/src/v/cluster_link/replication/partition_replicator.cc#L143) and then kept alive until the replication succeeds not.

>  is it possible the queue is destroyed before batches are consumed ?
> f we need to manage total memory usage maybe it would be better to have the per shard sidecar component that would account for all inflight batches memory usage ?

is the per shard thing only to prevent semaphore getting destroyed before units are returned or is there some other purpose?

mmaslankaprv commented on src/v/cluster_link/replication/deps_impl.cc:
```diff
@@ -0,0 +1,157 @@
+      !batches.empty(),
+      "Cannot replicate empty batch vector {}",
+      _partition->ntp());
+    chunked_vector<kafka::offset> expected_offsets;
```
nit: reserve

bharathv reviewed (commented):
thanks for the review, couple of clarifying questions.

bharathv committed:
cluster_link/replicator: fix stop order

Fully wait for the gate to close before stopping source. Source has a
semaphore (to be added in future commits) that will be destroyed as a
part of stop(), ensure all the units are returned first before calling
source::stop()

bharathv committed:
tests/utils: add a utility to produce a record batch

bharathv committed:
tests/utils: add a utility to consume records from partition

bharathv commented:
I'll try to add some e2e tests in ducktape using `ShadowLinkTestBase`, but otherwise this is ready for review.

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,218 @@
+        throw std::runtime_error(
+          ssx::sformat(
+            "Partition {} already registered with the consumer", tp));
+    }
```
quesiton: do we want to throw an exception or return an error code?

bharathv replied:
done.

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,218 @@
+    auto it = _partitions.find(tp);
+    if (it == _partitions.end()) {
+        return;
+    }
```
question: should this be an error if the tp isn't in the `_partitions` list?  Let the caller figure out next step

bharathv replied:
done.

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,218 @@
+    auto holder = _gate.hold();
+    auto it = _partitions.find(tp);
+    if (it == _partitions.end()) {
+        throw ss::abort_requested_exception{};
```
why is this an abort requested exception?

bharathv replied:
switched to error code.

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,218 @@
+    co_await _consumer->stop();
+    co_await std::move(f);
+    co_await ss::max_concurrent_for_each(
+      _partitions, 5, [](auto& p) { return p.second->stop(); });
```
nit: 5 is magic number

michael-redpanda replied:
should stop also clear out the content of `_partitions`?

oleiman commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_fixture.cc:
```diff
@@ -191,14 +206,25 @@ void basic_consumer_fixture::SetUp() {
       direct_consumer::configuration{
         .with_sessions = fetch_sessions_enabled{
           GetParam() == session_config::with_sessions}});
+    StartConsumer();
```
probably want to remove lines 204-208 yeah?

bharathv replied:
no? Because Start/StopConsumer() are only intended to control the start and stop of the created consumer (not its creation).

oleiman replied:
don't we call `make_consumer` in `StartConsumer`?

bharathv replied:
ah lol, right, sorry I need some ☕

oleiman commented on src/v/cluster_link/replication/deps.h:
```diff
@@ -31,20 +31,20 @@ class data_sink {
     virtual raft::replicate_stages replicate(
-      chunked_vector<model::record_batch> batches,
-      model::timeout_clock::duration timeout,
+      chunked_vector<::model::record_batch> batches,
```
in retrospect, I think having a module-scoped `namespace model` might be more trouble than it is worth.

bharathv committed:
cluster_link/replication/deps: add data_source_factory start/stop

..needed for later commits

bharathv committed:
tests/cluster_link: enable configs during rp init

No need to do it as a separate step.

bharathv commented:
Rebasing on `dev` shortly to include a fix in `ShadowLinkTestBase`.

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.h:
```diff
@@ -0,0 +1,95 @@
+    ss::future<>
+      process_fetched_data(chunked_vector<kafka::client::fetched_topic_data>);
+    // A ptr to allow std::move() without closing the queue gate.
+    chunked_hash_map<
```
would we benefit from making this a two level map ?

bharathv replied:
I don't think so because all the lookups are on `tp`.

mmaslankaprv commented on src/v/cluster_link/replication/mux_remote_consumer.h:
```diff
@@ -0,0 +1,95 @@
+ * be dynamically added or removed during runtime depending on if their queue is
+ * full.
+ */
+class mux_remote_consumer {
```
this is a perfect place to add metrics allowing us to see the throughput

mmaslankaprv commented on src/v/cluster_link/replication/deps_impl.cc:
```diff
@@ -0,0 +1,164 @@
+    }
+    if (_partition->term() == term) {
+        ssx::spawn_with_gate(_gate, [this, term] {
+            return _partition->raft()->step_down(
```
can you explain why do we need to stepdown here ?

bharathv replied:
added a comment

michael-redpanda commented on src/v/cluster_link/replication/deps_impl.cc:
```diff
@@ -0,0 +1,164 @@
+      "replicated: {}, attempting to replicate: {}",
+      _partition->ntp(),
+      _last_replicated_offset,
+      new_last_replicated);
```
question: should this be done before the call to `replicate`?

bharathv replied:
good catch, done.

michael-redpanda commented on src/v/cluster_link/link.cc:
```diff
@@ -56,19 +55,27 @@ link::link(
   , _config(std::move(config))
   , _cluster_connection(std::move(cluster_connection))
+  , _replication_mgr(
+      // todo: fix me
+      ss::default_scheduling_group(),
```
Not sure if you're based off of it ,but I did create a SG for cluster linking when I wired it into application.cc

bharathv replied:
can you point me to it, please? My git grep skills not helping me

michael-redpanda replied:
https://github.com/redpanda-data/redpanda/blob/bfd665d0b276c999cca4b25d60bd05aa852b5ad3/src/v/resource_mgmt/smp_groups.h#L51

bharathv replied:
these are different from scheduling_groups right? (needed for cross core invoke_on calls). We need to add a scheduling group to cpu_scheduling.h

michael-redpanda replied:
oh my bad ok

michael-redpanda replied:
yeah i got them confused in my head

michael-redpanda commented on src/v/cluster_link/link.cc:
```diff
@@ -168,6 +176,12 @@ ss::future<> link::handle_on_leadership_change(
+        _replication_mgr.start_replicator(ntp, *term);
+    } else {
+        _replication_mgr.stop_replicator(ntp, term);
+    }
```
I think you should check that the topic is a mirrored topic owned by the link before replication start?

bharathv replied:
good catch, done.

michael-redpanda commented on src/v/cluster_link/service.cc:
```diff
@@ -98,8 +108,36 @@ class default_link_factory : public link_factory {
+        cfg.isolation_level = ::model::isolation_level::read_committed;
+        cfg.max_buffered_bytes = 5_MiB;
+        cfg.max_buffered_elements = std::numeric_limits<size_t>::max();
+        cfg.with_sessions = kafka::client::fetch_sessions_enabled::yes;
```
not sure if you're based off of the right commit, but I do have some of these options in the model now

bharathv replied:
only a couple of them, right? min_bytes and max_fetch_size?  I don't see others in `connection_config`. Also we should make these dynamic (for tuning at runtime), can do in a follow up.

michael-redpanda replied:
yeah we can catch up on these in a follow up, np

michael-redpanda commented on src/v/redpanda/application.cc:
```diff
@@ -3147,6 +3148,10 @@ void application::start_runtime_services(
+              pm.register_factory<kafka::write_at_offset_stm_factory>(
+                storage.local().kvs(), model::offset_translator_batch_types());
+          };
       })
```
Does this require a restart then?  I only gate the API based off of the config.  Any reason not to just add the STM regardless of the config setting?

bharathv replied:
done

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -4368,7 +4368,7 @@ configuration::configuration()
       "development_enable_cluster_link",
       "Enable cluster linking.",
-      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      {.needs_restart = needs_restart::yes, .visibility = visibility::user},
```
Would like to not have to make this a need restart config if possible.

bharathv replied:
done.

michael-redpanda reviewed (commented):
This is 🔥 🔥 🔥

bharathv commented:
I have to rebase to fix a conflict.

bharathv commented:
aah... another conflict

bharathv commented:
Rebased to fix conflicts from fixture.h changes

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,247 @@
+}
+} // namespace cluster_link::replication
+
+auto fmt::formatter<cluster_link::replication::mux_remote_consumer::errc>::
+  format(
+    const cluster_link::replication::mux_remote_consumer::errc& e,
+    fmt::format_context& ctx) const -> decltype(ctx.out()) {
+    switch (e) {
+    case cluster_link::replication::mux_remote_consumer::errc::
+      partition_not_found:
+        return fmt::formatter<std::string_view>::format(
+          "partition_not_found", ctx);
+    case cluster_link::replication::mux_remote_consumer::errc::
+      partition_already_exists:
+        return fmt::formatter<std::string_view>::format(
+          "partition_already_exists", ctx);
+    }
+}
```
nit: you can use `format_as` for `enum` but this is fine

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -0,0 +1,247 @@
+    }
+}
+
+std::ostream& operator<<(
+  std::ostream& os,
+  const cluster_link::replication::mux_remote_consumer::errc& e) {
+    return os << fmt::format("{}", e);
+}
```
is this necessary if you have a `fmt::formatter`?

bharathv replied:
gtest formatters won't compile without this

ASSERT_EQ(foo...) << errc..

michael-redpanda commented on src/v/cluster_link/replication/partition_data_queue.cc:
```diff
@@ -0,0 +1,102 @@
+ss::future<partition_data_queue::fetch_data>
+partition_data_queue::fetch(ss::abort_source& as) {
+    auto holder = _gate.hold();
+    if (_waiter) {
+        throw std::runtime_error("Only one fetch can be in progress at a time");
+    }
```
question: is this an assertable event?

bharathv replied:
this was an assert originally but someone (Michal?) suggested better not to assert and instead throw.

michael-redpanda replied:
ack

michael-redpanda commented on src/v/cluster_link/replication/deps_impl.cc:
```diff
@@ -0,0 +1,200 @@
+ss::future<> remote_partition_source::stop() noexcept {
+    vlog(cllog.trace, "[{}] Stopping remote partition source", _tp);
+    auto f = _gate.close();
+    co_await _consumer.remove(_tp);
```
I think this can throw, right?  Do we want to handle that?

bharathv replied:
no? It returns an std::unexpected(..) in case of error.

michael-redpanda replied:
ah ok maybe it was a different method

michael-redpanda commented on src/v/cluster_link/replication/deps_impl.cc:
```diff
@@ -0,0 +1,200 @@
+          "[{}] Failed to reset remote partition source: {}",
+          _tp,
+          err);
+        return ss::make_exception_future<>(err);
```
maybe a dumb question, but why not just return the error?

bharathv replied:
that's a good question. The original API for the data source included a reset() function that was futurized and returned void (as defined in deps.h). After completing the implementation, it turned out that futurizing it wasn't actually necessary. now making that change requires changing a bunch of places, so I just left it as is, hoping no one would notice :)

michael-redpanda replied:
hahaha

michael-redpanda reviewed (commented):
looking good, just a few quesitons

dotnwat commented on src/v/cluster_link/link.cc:
```diff
@@ -90,7 +98,16 @@ ss::future<> link::stop() {
+              cllog.warn,
+              "Failed to stop task {}: {}",
+              t->name(),
+              res_f.get_exception());
```
FYI i don't think we should be doing this--correctness here depends on `vlog` always evaluating all of its parameters otherwise we get abanonded exceptional future exception here. it's similar to avoiding ever having side effects in an assert.

bharathv replied:
thats a good point, will fix in one of the followup PRs in this area.
